velox/vector/BaseVector.h

```diff
@@ -664,12 +664,17 @@ class BaseVector {
   virtual void prepareForReuse();
 
   // True if left and right are the same or if right is
-  // TypeKind::UNKNOWN.  ArrayVector copying may come across unknown
-  // type data for null-only content. Nulls can be transferred between
+  // TypeKind::UNKNOWN or left and right are TypeKind::VARCHAR and
+  // TypeKind::VARBINARY respectively. ArrayVector copying may come across
+  // unknown type data for null-only content. Nulls can be transferred between
   // two unknowns but values cannot be assigned into an unknown 'left'
   // from a not-unknown 'right'.
   static bool compatibleKind(TypeKind left, TypeKind right) {
-    return left == right || right == TypeKind::UNKNOWN;
+    // Vectors of VARCHAR and VARBINARY are compatible with each other.
+    bool varcharAndBinary =
+        (left == TypeKind::VARCHAR && right == TypeKind::VARBINARY) ||
```

Comment:
I am not 100% certain this is right, since Varchar also store additional metadata like asciiness etc , which wont get copied. I am assuming there are good historical reasons this isnt supported but I will validate. 

cc: @laithsakka .

Reply:
Thanks, Looking forward to your reply.